src/probabilities_estimators/symbolic/utils_spatial.jl

```diff
@@ -0,0 +1,80 @@
+# This file contains functions that are common to all spatial symbolic estimators.
+# --------------------------------------------------------------------------------
+# A convenience abstract type that makes common dispatch for pixel retrieval easier.
+abstract type SpatialProbEst{D, P} <: ProbabilitiesEstimator end
+
+# get stencil in the form of vectors of cartesian indices from either input type
+stencil_to_offsets(stencil::Vector{CartesianIndex{D}}) where D = stencil, D
+
+function stencil_to_offsets(stencil::NTuple{2, NTuple{D, T}}) where {D, T}
+    # get extent and lag from stencil
+    extent, lag = stencil
+    # generate a D-dimensional stencil
+    # start by generating a list of iterators for each dimension
+    iters = [0:lag[i]:extent[i]-1 for i in 1:D]
+    # then generate the stencil. We use an iterator product that we basically only reshape
+    # after that
+    stencil = CartesianIndex.(vcat(collect(Iterators.product(iters...))...))
+    return stencil, D
+end
+
+function stencil_to_offsets(stencil::Array{Int, D}) where D
+    # translate D-dim array into stencil of cartesian indices (of dimension D)
+    stencil = [idx - CartesianIndex(Tuple(ones(Int, D))) for idx in findall(Bool.(stencil))]
+    # subtract first coordinate from everything to get a stencil that contains (0,0)
+    stencil = [idx - stencil[1] for idx in stencil]
+    return stencil, D
+end
+
+"""
+    stencil_length(stencil::Vector{CartesianIndex{D}}) where D → length(stencil)
+    stencil_length(stencil::Array{Int, D}) where D → sum(stencil)
+    stencil_length(stencil::NTuple{2, NTuple{D, T}}) where {D, T} → prod(stencil[1])!
+
+Count the number of elements in the `stencil`.
+"""
+function stencil_length end
+stencil_length(stencil::Vector{CartesianIndex{D}}) where D = length(stencil)
+stencil_length(stencil::Array{Int, D}) where D = sum(stencil)
+stencil_length(stencil::NTuple{2, NTuple{D, T}}) where {D, T} = factorial(prod(stencil[1]))
+
+function preprocess_spatial(stencil, x::AbstractArray, periodic::Bool = true)
+    arraysize = size(x)
+    stencil, D = stencil_to_offsets(stencil)
+    @assert length(arraysize) == D "Indices and input array must match dimensionality!"
```

Comment:
Assertions are fine, but yes the alternative is 
```
length(arraysize) == D || error("...")
```
which honestly doesn't make much of a difference in neither source or what the user will see.

Reply:
Where did you learn this about assertions

Reply:
> I was recently told that assertions are not good practice for these kind of checks, but should rather be implemented as Errors/ Warnings because one can ignore AssertionErrors easily. But then again, I'm not too familiar with ErrorHandling in Julia, so maybe this is different here?
> Where did you learn this about assertions

[This](https://discourse.julialang.org/t/efficient-use-of-test-or-assert/75895/6) answer on JuliaDiscourse sums it up nicely: assertions in Julia can be turned off as an optimisation, so it shouldn't be user for input checking or similar things (which is what we're doing here).

> which honestly doesn't make much of a difference in neither source or what the user will see

```julia
length(arraysize) == D || error("...")
```

Using `error()` here will throw the most generic error possible, i.e. `ErrorException`. I think it is good practice to throw  error types that accurately reflect what type of condition was tested. This takes away the need for long error messages, and makes tests easier to write/read/understand. There's probably a build-in error that handles checks like these reasonably well.

Alternatively, one could write a custom error type like so:

```julia
type InputDimensionError{S<:AbstractString} <: Exception
    msg::S
end
Base.showerror(io::IO, e::InputDimensionError) = print(io, "InputDimensionError occured because ", e.msg, ".")
```
which could then be called as 

```julia
length(arraysize) == D || throw(InputDimensionError("Indices and input array must match dimensionality!")
```

This is arguably over-the-top for a simple input check like this, but can in principle be done.

Reply:
Standard error is fine. The most important infromation is in the string the error throws. In any case here the `DimensionMissmatch` error fits the most.